### ISSUE TITLE
MustacheEnvironmentCollector, InvalidPropertyException when JMustache is in the classpath

### ISSUE THREAD

**Issue by lfridael** [spring-projects/spring-boot#2610]
This configuration

```properties
# Necessary when Spring Security is in the classpath
shell.auth = simple
shell.auth.simple.user.name = name
shell.auth.simple.user.password = password
```

yields this exception, when JMustache autoconfiguration (which I think is new in 1.2.2) is also in the classpath:

```java
org.springframework.beans.InvalidPropertyException: Invalid property 'map[shell][auth][simple][user][password]' of bean class [org.springframework.boot.bind.RelaxedDataBinder$MapHolder]: Illegal attempt to get property 'map' threw exception; nested exception is org.springframework.beans.InvalidPropertyException: Invalid property 'map[shell][auth][simple][user][password]' of bean class [org.springframework.boot.bind.RelaxedDataBinder$MapHolder]: Property referenced in indexed property path 'map[shell][auth][simple][user][password]' is neither an array nor a List nor a Set nor a Map; returned value was [simple]
	at org.springframework.beans.BeanWrapperImpl.getPropertyValue(BeanWrapperImpl.java:853)
	at org.springframework.beans.BeanWrapperImpl.setPropertyValue(BeanWrapperImpl.java:948)
	at org.springframework.beans.BeanWrapperImpl.setPropertyValue(BeanWrapperImpl.java:927)
	at org.springframework.beans.AbstractPropertyAccessor.setPropertyValues(AbstractPropertyAccessor.java:95)
	at org.springframework.validation.DataBinder.applyPropertyValues(DataBinder.java:749)
	at org.springframework.validation.DataBinder.doBind(DataBinder.java:645)
	at org.springframework.boot.bind.RelaxedDataBinder.doBind(RelaxedDataBinder.java:119)
	at org.springframework.validation.DataBinder.bind(DataBinder.java:630)
	at org.springframework.boot.autoconfigure.mustache.MustacheEnvironmentCollector.setEnvironment(MustacheEnvironmentCollector.java:54)
	at org.springframework.boot.autoconfigure.mustache.MustacheAutoConfiguration.collector(MustacheAutoConfiguration.java:82)
	at org.springframework.boot.autoconfigure.mustache.MustacheAutoConfiguration.mustacheCompiler(MustacheAutoConfiguration.java:76)
	at org.springframework.boot.autoconfigure.mustache.MustacheAutoConfiguration$$EnhancerBySpringCGLIB$$69430051.CGLIB$mustacheCompiler$0(<generated>)
	at org.springframework.boot.autoconfigure.mustache.MustacheAutoConfiguration$$EnhancerBySpringCGLIB$$69430051$$FastClassBySpringCGLIB$$a923c792.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:228)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:309)
	at org.springframework.boot.autoconfigure.mustache.MustacheAutoConfiguration$$EnhancerBySpringCGLIB$$69430051.mustacheCompiler(<generated>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:162)
	... 48 more
```

**Comment by lfridael** [comment]
Workaround:

* Leave out `shell.auth = simple`
* Use this configuration instead:

```java
import org.springframework.boot.actuate.autoconfigure.ShellProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
class ShellConfig {

  @Bean
  public ShellProperties.CrshShellAuthenticationProperties simpleAuthenticationProperties() {
    return new ShellProperties.SimpleAuthenticationProperties();
  }

}
```

**Comment by snicoll** [comment]
Hello, I tried wit your config and I don't have an issue. Besides, I hardly understand how mustache would be related to the issue.

I created a sample app using https://start.spring.io with security, mustache, remote shell and the following config

```
shell.auth = simple
shell.auth.simple.user.name = name
shell.auth.simple.user.password = password
```

**Comment by lfridael** [comment]
@snicoll Thanks for your feedback. I've created a demo app and the issue does not occur there with this config, so this seems to be a problem with my application. Closing this issue.

**Comment by lfridael** [comment]
Sorry to bother you again. I've been able to reproduce the issue with a minimal example:
https://github.com/lfridael/spring-boot-mustache-issue

The problem is caused by a nested property in `application.properties`:
https://github.com/lfridael/spring-boot-mustache-issue/blob/master/src/main/resources/application.properties

Perhaps nested properties are not supported somehow? Anyway, the issue does not occur when mustache is not in the classpath.

**Comment by snicoll** [comment]
Cool, thanks! 

@dsyer the `MustacheEnvironmentCollector` looks like the culprit.

**Comment by dsyer** [comment]
It's actually the `RelaxedDataBinder`. I'll have a look.

**Comment by snicoll** [comment]
I think that probably fixed #2387 as well.

**Comment by snicoll** [comment]
Ironically, it does not fix #2387 and not even #2386 which is a bit surprising since I thought that was more or less what you did. I'll have a look.
